networking_cisco/plugins/cisco/db/l3/l3_router_appliance_gbp_db.py

```diff
@@ -0,0 +1,85 @@
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from networking_cisco.plugins.cisco.db.l3 import (
+    l3_router_appliance_db as l3_r_db)
+from networking_cisco.plugins.cisco.l3.drivers import driver_context
+
+from neutron.common import constants as l3_constants
+from neutron.common import exceptions as n_exc
+from neutron.i18n import _LW
+
+from oslo_config import cfg
+from oslo_log import log as logging
+
+FLOATINGIP_STATUS_ACTIVE = l3_constants.FLOATINGIP_STATUS_ACTIVE
+
+LOG = logging.getLogger(__name__)
+
+
+class L3RouterApplianceGbpDBMixin(l3_r_db.L3RouterApplianceDBMixin):
+    """Mixin class for Neutron's routing service using appliances with ACI.
+
+    This class should only be used for integrations with ACI using the
+    Group Based Policy (GBP) workflow. It's only difference from the parent
+    class is to specialize the create_floatingip call to support multiple
+    pools implemented by the GBP workflow.
+    """
+
+    def create_floatingip(self, context, floatingip,
+                          initial_status=FLOATINGIP_STATUS_ACTIVE):
+        """Specialization of parent class's create_floatingip.
+
+        This specialization of the parent class's create_floatingip
+        is needed for GBP workflow, as the pre-/post-commmit calls for
+        creating the floating IP must be peformed in a loop with the database
+        call in the grandparent class.
+        """
+
+        res = None
+        fip = floatingip['floatingip']
+        if not fip.get('subnet_id'):
+            # NOTE: default router type must be ASR1k
+            router_type_name = cfg.CONF.routing.default_router_type
+            driver = self._get_router_type_driver(context,
+                                                  router_type_name)
+            if driver:
+                fip_ctxt = driver_context.FloatingipContext(floatingip)
+                driver.create_floatingip_precommit(context, fip_ctxt)
+                nat_pool_list = getattr(context, 'nat_pool_list', [])
+                for nat_pool in nat_pool_list:
+                    if not nat_pool:
+                        continue
+                    fip['subnet_id'] = nat_pool['subnet_id']
+                    try:
+                        res = super(l3_r_db.L3RouterApplianceDBMixin,
+                                    self).create_floatingip(context,
+                                                            floatingip)
+                        router_ids = ([res['router_id']]
+                                      if res['router_id'] else [])
+                    except n_exc.IpAddressGenerationFailure as ex:
+                        LOG.warning(_LW("Floating allocation failed: %s"),
+                                    ex.message)
+                    if res:
+                        break
+        if not res:
+            info = super(l3_r_db.L3RouterApplianceDBMixin,
```

Comment:
Note that there are two ways to reach this code -- one is when a subnet_id is provided, and the other when the first allocation failed. I believe the intent is to still throw an exception if the allocation fails, which the bottom call would do. Also keep in mind that the first call is part of a loop -- it will log the message for each NAT pool allocation failure, and throw an exception if all of them fail.

One change that should be made here is the log message level. Bob and Sumit provided feedback on a similar patch here:

https://review.openstack.org/#/c/297307/2/gbpservice/neutron/services/grouppolicy/drivers/resource_mapping.py

I'll change the warning to an info.